src/nomad_simulations/properties/band_structure.py

```diff
@@ -0,0 +1,377 @@
+# limitations under the License.
+#
+
+import numpy as np
+from structlog.stdlib import BoundLogger
+from typing import Optional, Tuple
+import pint
+
+from nomad.metainfo import Quantity, Section, Context, SubSection
+
+from nomad_simulations.numerical_settings import KSpace
+from nomad_simulations.physical_property import PhysicalProperty
+from nomad_simulations.properties import ElectronicBandGap, FermiSurface
+from nomad_simulations.utils import get_sibling_section
+
+
+class BaseElectronicEigenvalues(PhysicalProperty):
+    """
+    A base section used to define basic quantities for the `ElectronicEigenvalues`, `FermiSurface`, and
+    `ElectronicBandStructure` properties. This section serves to define `FermiSurface` and without needing to specify
+    other quantities that appear in `ElectronicEigenvalues`
+    """
+
+    iri = ''
+
+    n_bands = Quantity(
+        type=np.int32,
+        description="""
+        Number of bands / eigenvalues.
+        """,
+    )
+
+    value = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Value of the electronic eigenvalues.
+        """,
+    )
+
+    def __init__(
+        self, m_def: Section = None, m_context: Context = None, **kwargs
+    ) -> None:
+        super().__init__(m_def, m_context, **kwargs)
+        # ! `n_bands` need to be set up during initialization of the class
+        self.rank = [kwargs.get('n_bands')]
+
+    def normalize(self, archive, logger) -> None:
+        super().normalize(archive, logger)
+
+
+class ElectronicEigenvalues(BaseElectronicEigenvalues):
+    """ """
+
+    iri = 'http://fairmat-nfdi.eu/taxonomy/ElectronicEigenvalues'
+
+    spin_channel = Quantity(
+        type=np.int32,
+        description="""
+        Spin channel of the corresponding electronic eigenvalues. It can take values of 0 or 1.
+        """,
+    )
+
+    occupation = Quantity(
+        type=np.float64,
+        shape=['*', 'n_bands'],
+        description="""
+        Occupation of the electronic eigenvalues. This is a number between 0 and 2, where 0 means
+        that the state is unoccupied and 2 means that the state is fully occupied. It is controlled
+        by the Fermi-Dirac distribution:
+
+            $ f(E) = 1 / (1 + exp((E - E_F) / kT)) $
+
+        The shape of this quantity is defined as `[KMesh.n_points, KMesh.dimensionality, n_bands]`, where `KMesh` is a `variable`.
+        """,
+    )
+
+    highest_occupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Highest occupied electronic eigenvalue. Together with `lowest_unoccupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    lowest_unoccupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Lowest unoccupied electronic eigenvalue. Together with `highest_occupied`, it defines the
+        electronic band gap.
+        """,
+    )
```

Comment:
I think it would be interesting to cover min/maxs of specific set of bands, e.g., around the Fermi level, higher energy bands separated by gaps, and equivalently, core bands separated by gaps.

I will add a question to see whether we can do this later along the road. Thanks for bringing it up.